pkg/fqdn/service/cell.go

```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package service
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/job"
+	"github.com/spf13/pflag"
+
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/fqdn/messagehandler"
+	"github.com/cilium/cilium/pkg/ipcache"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+// Cell provides the standalone DNS proxy gRPC server.
+// It is responsible for sending the DNS rules and IP cache updates
+// to the standalone DNS proxy. It also handles the DNS responses
+// from the standalone DNS proxy and updates the DNS rules and IP cache
+// accordingly. It receives the DNS rules during the endpoint regeneration
+// event and listens for ip cache updates from the ipcache.
+var Cell = cell.Module(
+	"sdp-grpc-server",
+	"Provides the standalone DNS proxy gRPC server",
+
+	cell.Config(defaultConfig),
+	cell.Provide(newServer),
+)
+
+type serverParams struct {
+	cell.In
+
+	Logger            *slog.Logger
+	EndpointManager   endpointmanager.EndpointManager
+	DNSRequestHandler messagehandler.DNSMessageHandler
+	IPCache           *ipcache.IPCache
+	JobGroup          job.Group
+	Config            FQDNConfig
+	DaemonConfig      *option.DaemonConfig
+}
+
+func newServer(params serverParams) *FQDNDataServer {
+	srv := NewServer(params.EndpointManager, params.DNSRequestHandler, params.Config.StandaloneDNSProxyServerPort, params.Logger)
+
+	if !params.Config.EnableStandaloneDNSProxy {
+		return srv
+	}
+
+	if !params.DaemonConfig.EnableL7Proxy {
+		srv.log.Error("Standalone DNS proxy requires L7 proxy to be enabled")
+		return srv
+	}
+
+	if params.DaemonConfig.ToFQDNsProxyPort == 0 || params.Config.StandaloneDNSProxyServerPort == 0 {
+		srv.log.Error("Standalone DNS proxy requires a valid port number to be set")
+		return srv
+	}
+
+	params.IPCache.AddListener(srv)
+
+	params.JobGroup.Add(job.OneShot("sdp-grpc-server", func(ctx context.Context, health cell.Health) error {
+		health.OK(fmt.Sprintf("Serving at %d", params.Config.StandaloneDNSProxyServerPort))
+		go srv.ListenAndServe()
```

Comment:
`ListenAndServe` returns an error which is ignored here, so either we pass in the health here and surface it that way or we make this a func which then updates health on error.

Also, the one shot job already runs in a goroutine, so I think you're spawning one here for the purpose of being able to call `Stop` on context cancellation?  but then if listening fails I don't get how we wouldn't silently not listen and block on the ctx done.

Reply:
Probably would make sense to just do `job.OneShot("sdp-grpc-server", srv.ListenAndServe)` and have all the logic in the same function.